simple-object-detection/train.js

```diff
@@ -0,0 +1,174 @@
+ *   `classLossMultiplier` to match the scale of the bounding-box loss
+ *   approximatey.
+ * - bounding-box loss, computed as the meanSquaredError between the
+ *   true and predicted bounding boxes.
+ * @param {*} yTrue
+ * @param {*} yPred
+ */
+function customLossFunction(yTrue, yPred) {
+  return tf.tidy(() => {
+    const batchSize = yTrue.shape[0];
+    const boundingBoxDims = yTrue.shape[1] - 1;
+
+    // Extract the shape-class portions of `yTrue` and `yPred`.
+    const classTrue = yTrue.slice([0, 0], [batchSize, 1]);
+    const classPred = tf.sigmoid(yPred.slice([0, 0], [batchSize, 1]));
+    const classLoss =
+        tf.metrics.binaryCrossentropy(classTrue, classPred).mean();
+
+    // Extract the bounding-box portions of `yTrue` and `yPred`.
+    const boundingBoxTrue = yTrue.slice([0, 1], [batchSize, boundingBoxDims]);
+    const boundingBoxPred = yPred.slice([0, 1], [batchSize, boundingBoxDims]);
+    const boundingBoxLoss =
+        tf.metrics.meanSquaredError(boundingBoxTrue, boundingBoxPred);
+
+    // Add the two losses to get the total loss. Note that `classLoss` is
+    // scaled by a factor to match bounding-box loss in scale approximately.
+    const totalLoss =
+        classLoss.mulStrict(classLossMultiplier).add(boundingBoxLoss);
+    return totalLoss;
+  });
+}
+
+/**
+ * Loads MobileNet and removes the top part.
+ * 
+ * Also gets handles to the layers that will be unfrozen during the fine-tuning
+ * phase of the training.
+ */
+async function loadDecapitatedMobilenet() {
+  const mobilenet = await tf.loadModel(
+      'https://storage.googleapis.com/tfjs-models/tfjs/mobilenet_v1_0.25_224/model.json');
+
+  // Return a model that outputs an internal activation.
+  const fineTuningLayers = [];
+  const layer = mobilenet.getLayer(topLayerName);
+  const decapitatedBase =
+      tf.model({inputs: mobilenet.inputs, outputs: layer.output});
+  // Freeze the model's layers.
+  for (const layer of decapitatedBase.layers) {
+    layer.trainable = false;
+    for (const groupName of topLayerGroupNames) {
+      if (layer.name.indexOf(groupName) === 0) {
+        fineTuningLayers.push(layer);
+        break;
+      }
+    }
+  }
+
+  tf.util.assert(
+      fineTuningLayers.length > 1,
+      `Did not find any layers that match the prefixes ${topLayerGroupNames}`);
+  return {decapitatedBase, fineTuningLayers};
+}
+
+/**
+ * Builds object-detection model from MobileNet.
+ */
+async function buildObjectDetectionModel() {
+  const {decapitatedBase, fineTuningLayers} = await loadDecapitatedMobilenet();
+
+  // Build the new head model.
+  const newHead = tf.sequential();
+  newHead.add(tf.layers.flatten(
+      {inputShape: decapitatedBase.outputs[0].shape.slice(1)}));
+  newHead.add(tf.layers.dense(
+      {units: 200, activation: 'relu', kernelInitializer: 'leCunNormal'}));
+  newHead.add(tf.layers.dense({units: 5, kernelInitializer: 'leCunNormal'}));
+  const newOutput = newHead.apply(decapitatedBase.outputs[0]);
```

Comment:
I'm not a fan of mixing the sequence and apply approachs in a single function.  But I'm not sure if 

const {decapitated Base, funeTuningLayers} = await ...
const newHead = buildHead(decipatedBase.outputs[0]);
const newModel = newHead.apply(decapitatedBazse.outputs[0]);
would be any better.

Thoughts?